### PR TITLE
feat: CloudBuild workspace blueprint

### PR DESCRIPTION
**Pull request by bharathkkb**
This blueprint creates an opinionated workflow for actuating Terraform using Cloud Build.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: build/int.cloudbuild.yaml [hunk 1]
```diff
@@ -93,6 +93,26 @@ steps:
       - verify-tfbuilder
   name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
   args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildBuilder --stage teardown --verbose']
+- id: init-tfworkpace
```
Comment (daniel-cit) [inline build/int.cloudbuild.yaml:96]:
```suggestion
- id: init-tfworkspace
```

File: build/int.cloudbuild.yaml [hunk 2]
```diff
@@ -93,6 +93,26 @@ steps:
...
+  waitFor:
+      - prepare
+  name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
+  args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildWorkspaceSimple --stage init --verbose']
+- id: apply-tfworkpace
+  waitFor:
+      - init-tfworkpace
+  name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
+  args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildWorkspaceSimple --stage apply --verbose']
+- id: verify-tfworkpace
```
Comment (daniel-cit) [inline build/int.cloudbuild.yaml:106]:
```suggestion
- id: verify-tfworkspace
```

File: build/int.cloudbuild.yaml [hunk 3]
```diff
@@ -93,6 +93,26 @@ steps:
...
+  waitFor:
+      - apply-tfworkpace
```
Comment (daniel-cit) [inline build/int.cloudbuild.yaml:108]:
```suggestion
      - apply-tfworkspace
```

File: build/int.cloudbuild.yaml [hunk 4]
```diff
@@ -93,6 +93,26 @@ steps:
...
+  name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
+  args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildWorkspaceSimple --stage verify --verbose']
+- id: teardown-tfworkpace
```
Comment (daniel-cit) [inline build/int.cloudbuild.yaml:111]:
```suggestion
- id: teardown-tfworkspace
```

File: build/int.cloudbuild.yaml [hunk 5]
```diff
@@ -93,6 +93,26 @@ steps:
...
+  waitFor:
+      - verify-tfworkpace
```
Comment (daniel-cit) [inline build/int.cloudbuild.yaml:113]:
```suggestion
      - verify-tfworkspace
```

File: examples/tf_cloudbuild_workspace_simple/files/backend.tf [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  backend "gcs" {
+    bucket = "tf-state-prod"
```
Comment (daniel-cit) [inline examples/tf_cloudbuild_workspace_simple/files/backend.tf:19]: do we need instructions to replace this fixed value ,`tf-state-prod`, with the bucket created in the seed project?

File: modules/tf_cloudbuild_workspace/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  cloudbuild_sa_roles     = { var.project_id = ["roles/compute.networkAdmin"] }
+}
+```
+
+Functional examples are included in the [examples](../../examples/) directory.
+
+## Resources created
+
+This module creates:
+- Two Cloud Build triggers with an inline build configuration for planning and applying Terraform configuration. Optionaly custom in repo build configs can be specified.
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/README.md:25]:
```suggestion
- Two Cloud Build triggers with an inline build configuration for planning and applying Terraform configuration. Optionally custom in repo build configs can be specified.
```

File: modules/tf_cloudbuild_workspace/buckets.tf [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  version = "~> 3.2"
+
+  name          = "${local.default_prefix}-build-state-${var.project_id}"
+  project_id    = var.project_id
+  location      = var.location
+  force_destroy = var.buckets_force_destroy
+}
+
+resource "google_storage_bucket_iam_member" "state_admin" {
+  bucket = module.state_bucket.bucket.self_link
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/buckets.tf:55]:
```suggestion
  bucket = local.state_bucket_self_link
```

File: modules/tf_cloudbuild_workspace/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+output "cloudbuild_sa" {
+  description = "SA used by Cloud Build triggers"
+  value       = local.cloudbuild_sa
+}
+
+output "state_bucket" {
+  description = "Bucket for storing TF state"
+  value       = module.state_bucket.bucket.self_link
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/outputs.tf:34]:
```suggestion
  value       = local.state_bucket_self_link 
```

File: modules/tf_cloudbuild_workspace/sa.tf [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+resource "google_service_account" "cb_sa" {
+  count        = var.cloudbuild_sa == "" ? 1 : 0
+  project      = var.project_id
+  account_id   = "tf-cb-${local.default_prefix}"
+  display_name = "SA for Terraform build trigger ${local.default_prefix}. Managed by Terraform."
+}
+
+# https://cloud.google.com/build/docs/securing-builds/configure-user-specified-service-accounts#permissions
+resource "google_project_iam_member" "cb_sa_roles" {
+  for_each = { for pr in local.cb_sa_roles_with_log : "${pr.project}/${pr.role}" => pr }
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/sa.tf:42]: @bharathkkb does it work if the projects are created in the same configuration?
https://www.terraform.io/language/meta-arguments/for_each#limitations-on-values-used-in-for_each

File: modules/tf_cloudbuild_workspace/variables.tf [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  default     = []
+}
+
+variable "buckets_force_destroy" {
+  description = "When deleting the bucket for storing CloudBuild logs/TF state, this boolean option will delete all contained objects. If false, Terraform will fail to delete buckets which contain objects."
+  type        = bool
+  default     = false
+}
+
+variable "substitutions" {
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/variables.tf:70]: if the user is providing the full set of substitution, how should the possible substitutions that are created inside the module
`_TF_SA_EMAIL`, `_STATE_BUCKET_NAME`, and `_LOGS_BUCKET_NAME` be handle?

File: build/int.cloudbuild.yaml (hunk 1 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: build/int.cloudbuild.yaml [hunk 12]
```diff
@@ -93,6 +93,26 @@ steps:
       - verify-tfbuilder
   name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
   args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildBuilder --stage teardown --verbose']
+- id: init-tfworkpace
+  waitFor:
+      - prepare
+  name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
+  args: ['/bin/bash', '-c', 'cft test run TestTFCloudBuildWorkspaceSimple --stage init --verbose']
+- id: apply-tfworkpace
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: build/int.cloudbuild.yaml (hunk 2 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: build/int.cloudbuild.yaml (hunk 3 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: build/int.cloudbuild.yaml (hunk 4 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: build/int.cloudbuild.yaml (hunk 5 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: done

File: modules/tf_cloudbuild_workspace/README.md (hunk 7 above)
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/README.md]: fixed

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
Checking for missing newline at end of file
Error: No newline at end of file ./examples/tf_cloudbuild_workspace_simple/files/.gitignore
 ```

File: examples/tf_cloudbuild_workspace_simple/files/backend.tf (hunk 6 above)
Comment (bharathkkb) [reply examples/tf_cloudbuild_workspace_simple/files/backend.tf:19]: For most use cases I am expecting this module to be used with inline config where we use partial backend config to override this. In SFB use a similar approach + subst with our custom build config that we ask users to copy over?

https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/e83d8ea8e60aa95e3ba53c18c8fc4bae22ea2198/modules/tf_cloudbuild_workspace/cb.tf#L37

File: modules/tf_cloudbuild_workspace/buckets.tf [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+resource "google_storage_bucket_iam_member" "log_admin" {
+  bucket = module.log_bucket.bucket.self_link
+  role   = "roles/storage.admin"
+  member = "serviceAccount:${local.cloudbuild_sa_email}"
+}
+
+# Custom bucket for storing TF state
+module "state_bucket" {
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/buckets.tf:45]: missed this, fixed

File: modules/tf_cloudbuild_workspace/buckets.tf (hunk 8 above)
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/buckets.tf]: done

File: modules/tf_cloudbuild_workspace/sa.tf (hunk 10 above)
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/sa.tf]: good catch, I ended up using static keyed maps to work around this in https://github.com/terraform-google-modules/terraform-google-bootstrap/pull/162/commits/cecd325706dabb88eb7f17de5924a8f37877bf11

File: modules/tf_cloudbuild_workspace/variables.tf (hunk 11 above)
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/variables.tf:73]: Added these as default substs merged with user provided.

File: modules/tf_cloudbuild_workspace/outputs.tf (hunk 9 above)
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/outputs.tf]: done

File: modules/tf_cloudbuild_workspace/cb.tf [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    "apply" = var.cloudbuild_apply_filename
+  }
+
+  # default prefix computed from repo name and dir if specified of form ${repo}-${dir?}-${plan/apply}
+  repo           = local.is_source_repo ? local.source_repo_name : local.gh_name
+  default_prefix = var.prefix != "" ? var.prefix : replace(var.tf_repo_dir != "" ? "${local.repo}-${var.tf_repo_dir}" : local.repo, "/", "-")
+
+  # default substitutions
+  default_subst = {
+    "_TF_SA_EMAIL"       = local.cloudbuild_sa,
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/cb.tf:56]:
```suggestion
    "_TF_SA_EMAIL"       = local.cloudbuild_sa_email 
```

File: modules/tf_cloudbuild_workspace/README.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+Basic usage of this module is as follows:
+
+```hcl
+module "tf-net-workspace" {
+  source  = "terraform-google-modules/bootstrap/google//modules/tf_cloudbuild_workspace"
+  version = "~> 6.1"
+
+  project_id              = var.project_id
+  tf_repo_uri             = "https://github.com/org/tf-config-repo"
+  cloudbuild_sa_roles     = { var.project_id = ["roles/compute.networkAdmin"] }
```
Comment (daniel-cit) [inline modules/tf_cloudbuild_workspace/README.md:16]:
```suggestion
  project_id          = var.project_id
  tf_repo_uri         = "https://github.com/org/tf-config-repo"
  cloudbuild_sa_roles = { var.project_id = ["roles/compute.networkAdmin"] }
```

**Review by daniel-cit — CHANGES_REQUESTED** [review]
Thanks for the contribution.

There are some issues to address:
 - errors in cases of `count`/`for_each` in the context when the inputs used are create in the same configuration
 - adjustments in the README.md
 -  adjustments in  `cloudbuild_sa` usage

File: modules/tf_cloudbuild_workspace/sa.tf [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+          project_id = pr.project_id, role = role
+        }
+      }
+    ]...
+  )
+}
+
+
+resource "google_service_account" "cb_sa" {
+  count        = var.cloudbuild_sa == "" ? 1 : 0
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/sa.tf]: exposed create as a flag

File: modules/tf_cloudbuild_workspace/buckets.tf [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  source  = "terraform-google-modules/cloud-storage/google//modules/simple_bucket"
+  version = "~> 3.2"
+  count   = var.state_bucket_self_link != "" ? 0 : 1
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/buckets.tf]: exposed create as a flag

File: modules/tf_cloudbuild_workspace/sa.tf [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+# cross project impersonation if a custom CB SA is specified from a different project
+resource "google_service_account_iam_member" "cb_sa_self" {
+  for_each           = local.diff_sa_project ? toset(["roles/iam.serviceAccountUser", "roles/iam.serviceAccountTokenCreator"]) : []
+  service_account_id = local.cloudbuild_sa
+  role               = each.value
+  member             = "serviceAccount:${local.cloudbuild_sa_email}"
+}
+
+data "google_project" "cloudbuild_project" {
+  count      = local.diff_sa_project ? 1 : 0
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/sa.tf]: exposed diff_sa_project as a flag

File: modules/tf_cloudbuild_workspace/variables.tf [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+variable "location" {
+  description = "Location for build logs/state bucket"
+  type        = string
+  default     = "us-central1"
+}
+
+variable "cloudbuild_sa" {
+  description = "Custom SA email to be used by the CloudBuild trigger. Defaults to being created if empty."
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/variables.tf]: specified that it has to be id with format

File: modules/tf_cloudbuild_workspace/README.md [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+|------|-------------|
+| cloudbuild\_apply\_trigger\_id | Trigger used for running TF apply |
+| cloudbuild\_plan\_trigger\_id | Trigger used for running TF plan |
+| cloudbuild\_sa | SA used by Cloud Build triggers |
+| logs\_bucket | Bucket for storing TF logs/plans |
+| state\_bucket | Bucket for storing TF state |
+
+<!-- END OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+
+## Requirements
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/README.md:72]: added notes

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: modules/tf_cloudbuild_workspace/sa.tf [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+# https://cloud.google.com/build/docs/securing-builds/configure-user-specified-service-accounts#permissions
+resource "google_project_iam_member" "cb_sa_logging" {
+  project = var.project_id
+  role    = "roles/logging.logWriter"
+  member  = "serviceAccount:${local.cloudbuild_sa_email}"
+}
+
+resource "google_project_iam_member" "cb_sa_roles" {
+  for_each = local.cb_sa_roles_expand
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/sa.tf:50]: I was unable to repro this, I tried not specifying `cloudbuild_sa_roles` and setting it explicitly to empty map but seemed to work in both cases.

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: modules/tf_cloudbuild_workspace/sa.tf (hunk 21 above)
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/tf_cloudbuild_workspace/sa.tf:50]: Updated module version constraint to 0.15 which has the fix for https://github.com/hashicorp/terraform/issues/26268

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@bharathkkb
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Review by daniel-cit — APPROVED** [review]
LGTM

**Pull request merged**